GVFS/GVFS.Common/Git/GitSsl.cs

```diff
@@ -0,0 +1,171 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Text.RegularExpressions;
+using GVFS.Common.Tracing;
+
+namespace GVFS.Common.Git
+{
+    public class GitSsl
+    {
+        public readonly string CertificatePathOrSubjectCommonName;
+        public readonly bool IsCertificatePasswordProtected;
+        public readonly bool ShouldVerify;
+
+        public GitSsl()
+        {
+            this.CertificatePathOrSubjectCommonName = null;
+            this.IsCertificatePasswordProtected = false;
+            this.ShouldVerify = true;
+        }
+
+        public GitSsl(IDictionary<string, GitConfigSetting> configSettings) : this()
+        {
+            if (configSettings != null)
+            {
+                if (configSettings.TryGetValue(GitConfigSetting.HttpSslCert, out GitConfigSetting sslCerts))
+                {
+                    this.CertificatePathOrSubjectCommonName = sslCerts.Values.Single();
+                }
+
+                if (configSettings.TryGetValue(GitConfigSetting.HttpSslCertPasswordProtected, out GitConfigSetting isSslCertPasswordProtected))
+                {
+                    this.IsCertificatePasswordProtected = isSslCertPasswordProtected.Values.Select(bool.Parse).Single();
+                }
+
+                if (configSettings.TryGetValue(GitConfigSetting.HttpSslVerify, out GitConfigSetting sslVerify))
+                {
+                    this.ShouldVerify = sslVerify.Values.Select(bool.Parse).Single();
+                }
+            }
+        }
+
+        public string GetCertificatePassword(ITracer tracer, GitProcess git)
+        {
+            if (git.TryGetCertificatePassword(tracer, this.CertificatePathOrSubjectCommonName, out string password, out string error))
+            {
+                return password;
+            }
+
+            return null;
+        }
+
+        public X509Certificate2 GetCertificate(ITracer tracer, string certificatePassword, bool onlyLoadValidCertificateFromStore)
+        {
+            EventMetadata metadata = new EventMetadata
+            {
+                { "CertificatePathOrSubjectCommonName", this.CertificatePathOrSubjectCommonName },

```

Comment:
Use `nameof(this.CertificatePathOrSubjectCommonName)` to avoid the string getting out of sync with the property name.

Reply:
It depends - do we want to keep the logs stable, regardless of potential renames of variables? If so, having a hardcoded string (which diverge from the variable name in the future) would give us that.

Reply:
Yes that is true but if it gets renamed to something extremely different like `CertificateExpirationDate` the name would not match the data so there is a balance there.  I try to use `nameof` by default if the name would just be the name of the property then if the property name ever changes you can decide if it better to keep the old name or allow it to change.  The code does use both hard coded strings and `nameof` so there aren't any hard and fast rules here so I'll leave it up to you.